enhancements/microshift/advertise-version-details.md

```diff
@@ -0,0 +1,303 @@
+information, we will keep the contents of the ConfigMap brief and
+specify the data key names in this enhancement:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: version
+  namespace: openshift
+data:
+  product: ""
+  major: ""
+  minor: ""
+  version: ""
+```
+
+- `product` should hold a representation of the OpenShift variant
+  name. For example, `OCP` for OpenShift in standalone or external
+  control-plane configurations of all topologies and `MicroShift` for
+  MicroShift instances.
+- `major` should hold the major version number of the product as a
+  string. For example, for `4.12.0` the value would be `"4"`.
+- `minor` should hold the minor version number of the product as a
+  string. For example, for `4.12.0` the value would be `"12"`.
+- `version` should hold the complete version of the product as a
+  string. For example, for `4.12.0` the value would be `"4.12.0"`. The
+  version string may also include other qualifiers in nightly,
+  sprintly, CI, or hot-fix builds. For example,
+  `4.11.0-0.nightly-2022-06-23-153912`.
+
+The `version` string is meant for display in user interface
+situations.  The `major` and `minor` values are included separately so
+that API users do not have to parse the `version` string before
+applying rules based on the values.
+
+### Workflow Description
+
+#### Determining the Version of a Cluster via the API
+
+API clients can read the `openshift/version` ConfigMap contents in the
+normal way.
+
+#### Determining the Version of a Cluster via the CLI
+
+The `oc login` command reports the `message` value of the
+`openshift/motd` ConfigMap when a user successfully authenticates. The
+command should be extended to include the `version` string of the
+`openshift/version` ConfigMap, when it is present.
+
+The `oc version` command reports version details today. Some of the
+information it reports comes from the ClusterVersion resource, which
+is not visible to all users. The command should be updated to look at
+the `openshift/version` ConfigMap when it is present and the
+ClusterVersion resource cannot be read.
+
+#### Publishing the Version in MicroShift
+
+MicroShift will update the `openshift/version` ConfigMap contents on
+startup using a value compiled into the main binary. Upgrading
+MicroShift requires restarting the service, so there is no need to
+continue reconciling the contents of the ConfigMap.
+
+#### Publishing the Version in OCP
+
+OpenShift running in standalone or external control-plane
+configurations may be upgraded dynamically in place without loss of
+API connectivity. The cluster-version-operator will therefore update
+the contents of the ConfigMap as the version of the cluster
+changes. During an upgrade, we want to report the current version as
+the one from which the cluster is upgrading and not expose that an
+upgrade is in progress. The ConfigMap should therefore always contain
+the value of the most recent version marked as `Completed` in the
```

Comment:
APIs are often added during 4.y -> 4.(y+1) minor version updates.  But sometimes [APIs are removed][1].  And then the other way around, in the event of folks rolling back from 4.(y+1) to 4.y, APIS could be both added and removed (although [OCP does not support rollbacks today][2]).  Minor version updates shouldn't be all that common, but it might be worth a paragraph under _Version Skew Strategy_ that says:

> There's a risk, but folks using this ConfigMap to drive choices via lookup tables are opting in to that skew risk, so they can deal with the consequences of their indirect approach.  And minor updates that bring in new APIs and remove deprecated APIs are infrequent and short, so even consumers bit by the skew window should be able to recover after the update completes.

Or something?

[1]: https://docs.openshift.com/container-platform/4.9/release_notes/ocp-4-9-release-notes.html#ocp-4-9-admin-ack-upgrading
[2]: https://github.com/openshift/openshift-docs/blame/bf973ccd31101cf54b6639df391f82c09ba421cf/updating/updating-cluster-cli.adoc#L30

Reply:
This is an issue when APIs are removed. But when we remove API we have extra layers (admin-ack) to protect users. Also I am not concerned by minor rollbacks. That's not something going to work for us in near future.

Reply:
Again, we should be clear that this version should not be used as decision maker, this information is purely informational. That's what I was talking about during our arch call last week.

Reply:
One of the big reasons for exposing the information in the first place is to let users, installers like Helm charts, etc. make decisions based on that information. We can recommend other/better approaches in specific cases where that's possible, but it isn't always possible (that's the point) and even when it is possible it's impractical for us to require purity of implementation in every corner of the ecosystem.